src/main/java/com/google/sps/data/DeliverySystem.java

```diff
@@ -26,16 +26,74 @@
 import java.util.ArrayList;
 import java.util.Date;
 import java.util.HashMap;
+import java.util.List;
 
 /** 
  * Class that assigns orders and creates delivery journeys for delivery slots.
  */
 public class DeliverySystem {
+  private final Integer MAX_WAYPOINTS = 25;
+  private PathFinder pathFinder;
+
+  public DeliverySystem(PathFinder pathFinder) {
+    this.pathFinder = pathFinder;
+  }
+
   /**
    * Finds the order in which points must be visited such that the journey lasts minimum time and it
    * delivers all orders assigned in findOrdersForDeliveryRequest.
    */
-  public ArrayList<Point> findOptimalDeliveryJourneyForDeliverySlot(DeliverySlot deliverySlot) throws ApiException, IOException, InterruptedException {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+  public ArrayList<Point> findOptimalDeliveryJourneyForDeliverySlot(DeliverySlot deliverySlot) throws ApiException, BadRequestException, DataNotFoundException, IOException, InterruptedException {
+    Journey journey = findOrdersForDeliveryRequest(deliverySlot);
+    return journey.findOptimalOrderForWaypoints();
+  }
+
+  /**
+   * Assigns a set of orders for the delivery request such that they fit the [startTime, endTime]
+   * timeslot.
+   */
+  public Journey findOrdersForDeliveryRequest(DeliverySlot deliverySlot)  throws ApiException, BadRequestException, DataNotFoundException, IOException, InterruptedException {
+    Point startPoint = deliverySlot.getStartPoint();
+    Journey journey = new Journey(startPoint, pathFinder);
+    OrderHandler orderHandler = new OrderHandler(pathFinder);
+     // Get unassigned orders from datastore which are in the area of the starting point.
+    List<Entity> orders = orderHandler.getAvailableOrders(startPoint.getArea());
+    for (Entity order : orders) {
+      if (journey.getNumberOfWaypoints() >= MAX_WAYPOINTS) {
+        break;
+      }
+
+      LibraryPoint library = new LibraryPoint((double) order.getProperty(OrderHandler.OrderProperty.LIBRARY_LAT.label), 
+        (double) order.getProperty(OrderHandler.OrderProperty.LIBRARY_LNG.label),
+        (int) order.getProperty(OrderHandler.OrderProperty.LIBRARY_ID.label));
+      Point recipient = new Point((double) order.getProperty(OrderHandler.OrderProperty.RECIPIENT_LAT.label), 
+        (double) order.getProperty(OrderHandler.OrderProperty.RECIPIENT_LNG.label));
+      
+      // Check if the library is already a waypoint in the journey, and add it in case it's not.
+      boolean libraryIsWaypoint = !journey.addPointToWaypoints(library);
+      // Check if the recipient is already a waypoint in the journey, and add it in case it's not.
+      boolean recipientIsWaypoint = !journey.addPointToWaypoints(recipient);
+
+      if (journey.getNumberOfWaypoints() > MAX_WAYPOINTS || !journey.findJourneyForTimeslot(deliverySlot)) {
```

Comment:
The problem with adding the points first is that libraries and the addresses of recipients are different, e.g. trimming waypoints to fit the delivery slot requirements might result in a set of waypoints containing only libraries.
Another example would be if:
recipient1 orders from library1 and library2
recipient2 orders from library3
The trimmed set can end up with recipient2, library1, library2, which is not valid.

Reply:
Yeah, after trimming waypoints, you would need to remove any library waypoints from which the courier suddenly has no books to pick up for (if you removed a recipient1, there's nothing to pick up from library1 and library2).

But yeah, you're right that it would then be theoretically possible that a user who lives near many libraries would not get any deliveries (or only library waypoints which would then get removed) if they have a very small timeslot.